File: javaSpringShip/src/main/java/pl/battleships/javaspringship/service/GameServiceImpl.java
package pl.battleships.javaspringship.service;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import pl.battleships.api.dto.GameDto;
import pl.battleships.api.dto.GameStatusDto;
import pl.battleships.api.dto.PositionDto;
import pl.battleships.api.dto.ShotStatusDto;
import pl.battleships.core.api.BattleshipGame;
import pl.battleships.core.api.HistoryProvider;
import pl.battleships.javaspringship.mapper.GameModelMapper;

import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@RequiredArgsConstructor
public class GameServiceImpl implements GameService {

    private final BattleshipGame battleshipGame;
    private final GameModelMapper mapper;
    private final HistoryProvider historyProvider;

    @Override
    public void joinTheGame(GameDto game) {
        battleshipGame.start(game.getId(), game.getSize(), game.getFirstShotIsYours());
    }

    @Override
    public ShotStatusDto opponentShot(String gameId, PositionDto position) {
        return mapper.convert(
                battleshipGame.opponentShot(gameId, mapper.convert(position))
        );
    }

    @Override
    public List<PositionDto> getAllShots(String id) {
        return historyProvider.getAllShots(id).stream().map(mapper::convert).collect(Collectors.toList());
    }

    @Override
    public GameStatusDto getGameStatus(String gameId) {
        return mapper.convert(battleshipGame.getGameStatus(gameId));
    }
}

File: kotlinSpringShip/src/main/kotlin/pl/battleships/kotlinspringship/service/GameServiceImpl.kt
package pl.battleships.kotlinspringship.service

import pl.battleships.api.dto.GameDto
import pl.battleships.api.dto.GameStatusDto
import pl.battleships.api.dto.PositionDto
import pl.battleships.api.dto.ShotStatusDto
import pl.battleships.core.api.BattleshipGame
import pl.battleships.core.api.HistoryProvider
import pl.battleships.kotlinspringship.extension.toGameStatusDto
import pl.battleships.kotlinspringship.extension.toPosition
import pl.battleships.kotlinspringship.extension.toPositionDto
import pl.battleships.kotlinspringship.extension.toShotStatusDto

class GameServiceImpl(
    private val battleshipGame: BattleshipGame,
    private val historyProvider: HistoryProvider, // trailing comma <3
) : GameService {

    override fun joinTheGame(game: GameDto) {
        battleshipGame.start(game.id, game.propertySize, game.firstShotIsYours)
    }

    override fun opponentShot(gameId: String, position: PositionDto): ShotStatusDto {
        return battleshipGame.opponentShot(gameId, position.toPosition()).toShotStatusDto()
    }

    override fun getAllShots(id: String): List<PositionDto> {
        return historyProvider.getAllShots(id).map { it.toPositionDto() }
    }

    override fun getGameStatus(gameId: String): GameStatusDto {
        return battleshipGame.getGameStatus(gameId).toGameStatusDto()
    }
}